src/core/server/index.ts

```diff
@@ -481,26 +481,34 @@ export { TelemetryCounterType } from './analytics';
  * @public
  */
 export interface RequestHandlerContext {
-  core: {
-    savedObjects: {
-      client: SavedObjectsClientContract;
-      typeRegistry: ISavedObjectTypeRegistry;
-      getClient: (options?: SavedObjectsClientProviderOptions) => SavedObjectsClientContract;
-      getExporter: (client: SavedObjectsClientContract) => ISavedObjectsExporter;
-      getImporter: (client: SavedObjectsClientContract) => ISavedObjectsImporter;
-    };
-    elasticsearch: {
-      client: IScopedClusterClient;
-    };
-    uiSettings: {
-      client: IUiSettingsClient;
-    };
-    deprecations: {
-      client: DeprecationsClient;
-    };
+  core: Promise<CoreRequestHandlerContext>;
+}
+
+/** @public */
+export interface CoreRequestHandlerContext {
+  savedObjects: {
+    client: SavedObjectsClientContract;
+    typeRegistry: ISavedObjectTypeRegistry;
+    getClient: (options?: SavedObjectsClientProviderOptions) => SavedObjectsClientContract;
+    getExporter: (client: SavedObjectsClientContract) => ISavedObjectsExporter;
+    getImporter: (client: SavedObjectsClientContract) => ISavedObjectsImporter;
+  };
+  elasticsearch: {
+    client: IScopedClusterClient;
+  };
+  uiSettings: {
+    client: IUiSettingsClient;
+  };
+  deprecations: {
+    client: DeprecationsClient;
   };
 }
 
+/** @public */
+export type CustomRequestHandlerContext<T> = RequestHandlerContext & {
+  [Key in keyof T]: T[Key] extends Promise<unknown> ? T[Key] : Promise<T[Key]>;
+};
```

Comment:
Utility type for plugins to define their own context types (previously it was done by extending `RequestHandlerContext`)